### DISCUSSION TITLE
Custom DataSources trough Extensibility Library

### DISCUSSION THREAD

**Discussion by edarroudi** [microsoft-search/pnp-modern-search#1471] [Ideas]
Hi,

would it be possible to extend the **IExtensibilityLibrary** on a method: **getCustomDataSourceProviders**() that extends the available datasources? My client wants to be able to display search results from some REST APIs where we don't want / can't use the Microsoft Search external item interace (Let's say i want to show a result from Wikipedia and don't want to create a index of it ;-)).

Then in loadExtensions we cann concat availableDataSourceDefinitions..

Of course SearchResultWebpart method getDataSourceInstance has to be changed somehow as well as far as i understand it.

Should i give it a try or is this a no go from your perspective due to security reasons or ...

**Comment by edarroudi** [comment]
I understand that i can just create a custom spfx webpart and do it all manually as well, i just think that this could be useful in many ways...

**Reply by FranckyC** [reply]
Hi @edarroudi, this feature already exist but not in the open source version. See https://aequos-solutions.github.io/modern-data-visualizer/extensibility/custom_data_source/ for more information.

**Reply by edarroudi** [reply]
Hi @FranckyC , 

ahh good to know there was another package with more features. I'll tell the company i'm working for about this.
I think they would be happy to buy the solution but i see some security concerns for them:

a) the code is not open source
b) there seems to be no possiblity to restrict the datasources the user can configure

Would it be possible to have a security audit on the code?

**Reply by FranckyC** [reply]
This solution shares the same code foundation as the PnP Modern Search so you can do a security audit on it. Also, you can't restrict the data sources user can configure. However I'm sure sure about the requirement here.

**Reply by edarroudi** [reply]
That won't be enough for the company, they would need the complete Code for the audit - i hope you understand this as an IT guy :) . Regarding the restriction: The fear that a "common" user configures an additional REST call to a server where let's say all search requests are protocolled. Just an example of a possible security & data compliance issue.

**Reply by FranckyC** [reply]
Hi @edarroudi, unfortunately we can't share our code directly to oher companies. If an user can do what you said, it first mean he has access to the page as contributor and already have special permissions. Also, you could test this specific scenario with the solution directly as we offer a free trial for 14 days.

The component code is just a wrapper over SPFx HttpClient/MSGraphClient with no additional checks.
